### PR TITLE
gcc-12 always-false in fread.c

### PR DESCRIPTION
**Pull request by mattdowle**
With gcc 12.1 as displayed on CRAN currently
```
Version: 1.14.2
Flags: --no-tests
Check: whether package can be installed
Result: WARN
    Found the following significant warnings:
     fread.c:1278:38: warning: the comparison will always evaluate as ‘false’ for the pointer operand
                               in ‘nastr + 8’ must not be NULL [-Waddress]
Flavor: r-devel-linux-x86_64-debian-gcc

### CODE REVIEW

**Comment by jangorecki** [comment]
There was an issue and PR already to resolve this
https://github.com/Rdatatable/data.table/issues/5377

**Comment by mattdowle** [comment]
NB: turning on `-Waddress` in `cc.R` with gcc 11.2 doesn't detect it. I tried adding `-Waddress` to cc.R so that we wouldn't need gcc 12 to detect it. `-Wall` is already on in `cc.R` so I guess that `-Wall` already includes `-Waddress` but that something has improved in gcc 12 that now detects this; i.e. it's not just that `-Waddress` is now included by -Wall in gcc 12.

File: src/fread.c [hunk 1]
```diff
@@ -1284,25 +1284,22 @@ int freadMain(freadMainArgs _args) {
...
+      const char *ch = *nastr;
+      size_t nchar = strlen(ch);
+      if (isspace(ch[0]) || isspace(ch[nchar-1]))
+        STOP(_("freadMain: NAstring <<%s>> has whitespace at the beginning or end"), ch);
+      if (strcmp(ch,"T")==0    || strcmp(ch,"F")==0 ||
+          strcmp(ch,"TRUE")==0 || strcmp(ch,"FALSE")==0 ||
+          strcmp(ch,"True")==0 || strcmp(ch,"False")==0)
+        STOP(_("freadMain: NAstring <<%s>> is recognized as type boolean, this is not permitted."), ch);
+      if ((strcmp(ch,"1")==0 || strcmp(ch,"0")==0) && args.logical01)
+        STOP(_("freadMain: NAstring <<%s>> and logical01=%s, this is not permitted."), ch, args.logical01 ? "TRUE" : "FALSE");
```
Comment (MichaelChirico) [inline src/fread.c:1297]: slightly confused by the templating in the string here am I missing something? in the error case logical01 is always TRUE?

File: src/fread.c [hunk 2]
```diff
@@ -1284,25 +1284,22 @@ int freadMain(freadMainArgs _args) {
...
+      char *end;
+      errno = 0;
+      (void)strtod(ch, &end);  // careful not to let "" get to here (see continue above) as strtod considers "" numeric
```
Comment (MichaelChirico) [inline src/fread.c:1300]: is "continue" referring to somewhere else in the file?

File: src/fread.c (hunk 1 above)
Comment (mattdowle) [reply src/fread.c:1297]: good spot. agree logical01=TRUE could be in the string and the ternary isn't needed.

File: src/fread.c (hunk 2 above)
Comment (mattdowle) [reply src/fread.c:1300]: indeed. good spot.

**Pull request merged**